### PR TITLE
drivers: sensor: bmp388: add OSR/ODR timing validation

### PR DESCRIPTION
**Pull request by matteoscordino**
This is a fresh attempt at getting this feature merged, after I closed a previous PR, #53025, that got messed up. This new PR already addresses the comments from #53025, the only reason this was re-submitted is a rebase that went wrong on the old one.

The BMP3xx won't start up correctly if the ODR/OSR settings are not consistent (the IC won't exit sleep mode when asked to go normal), and without this patch the driver is not detecting it; the init completes and the sensor looks "active", but no data is sampled. In triggered mode you will get no trigger, in polling mode you will loop forever doing I2C transactions to read the status register waiting for DRDY.

Even worse, if `CONFIG_PM_DEVICE` is enabled, the state of the IC would get out of sync with the state known by the PM logic.

So, consistency checking of the OSR/ODR setting is not a purely formal check.

An alternative (and simpler) approach would be to read back the power mode register after init and ODR/OSR setting, but that has a few significant drawbacks:

- it just says that init failed, not WHY
- it would add an I2C transaction to each attribute setting (assuming that dynamic setting is enabled)
- again for dynamic attribute setting, detecting the error after the fact would leave the IC in an inconsistent state, and require some form of recovery
- it would add an I2C transaction to init (not very significant, as it's a one-off)
- principle of least surprise: checking this is what the official Bosch driver does, so people coming from other platforms expect it

### CODE REVIEW

**Comment by teburd** [comment]
@teburd @MaureenHelm @yperess lets take a look at this

File: drivers/sensor/bmp388/bmp388.c [hunk 1]
```diff
@@ -207,6 +207,102 @@ int bmp388_reg_field_update(const struct device *dev,
...
+  uint32_t pow_output = 1;
+
+  while (power != 0)
+  {
+    pow_output = base * pow_output;
+    power--;
+  }
+
+  return pow_output;
+}
```
Comment (MaureenHelm) [inline drivers/sensor/bmp388/bmp388.c]: It looks like this function is always called with base = 2, so why not use a left shift instead?

File: drivers/sensor/bmp388/bmp388.c [hunk 2]
```diff
@@ -207,6 +207,102 @@ int bmp388_reg_field_update(const struct device *dev,
...
+   * pressure measurement duration 
+   */
+  meas_t += meas_t_p;
+  LOG_DBG("total meas_t: %uus", meas_t);
+  LOG_DBG("total odr_t: %uus", odr_period_lut[odr]);
+
+  if (meas_t > odr_period_lut[odr])
+  {
+	/* OSR and ODR settings are not proper */
+    return -EINVAL;
```
Comment (MaureenHelm) [inline drivers/sensor/bmp388/bmp388.c]: Suggest adding a LOG_ERR before returning

File: drivers/sensor/bmp388/bmp388.c [hunk 3]
```diff
@@ -240,6 +336,14 @@ static int bmp388_attr_set_odr(const struct device *dev,
 		return odr;
 	}
 
+	#if defined (CONFIG_BMP388_VALIDATE_TIMINGS)
+	/* check that the new ODR would be consistent with the current OSRs */
+	if ( 0 > bmp388_validate_osr_and_odr_settings(odr, data->osr_pressure, data->osr_temp))
```
Comment (MaureenHelm) [inline drivers/sensor/bmp388/bmp388.c]:
```suggestion
	if (bmp388_validate_osr_and_odr_settings(odr, data->osr_pressure, data->osr_temp) < 0)
```
Please do this throughout

File: drivers/sensor/bmp388/bmp388.c [hunk 4]
```diff
@@ -268,23 +372,43 @@ static int bmp388_attr_set_oversampling(const struct device *dev,
...
+
 	if (chan == SENSOR_CHAN_PRESS) {
 		pos = BMP388_OSR_PRESSURE_POS;
 		mask = BMP388_OSR_PRESSURE_MASK;
+		#if defined (CONFIG_BMP388_VALIDATE_TIMINGS)
+		/* check that the new press OSR would be consistent with the current 
+		 * ODR and the current temp OSR */
+		if ( 0 > bmp388_validate_osr_and_odr_settings(data->odr, 
+														reg_val, 
+														data->osr_temp))
```
Comment (MaureenHelm) [inline drivers/sensor/bmp388/bmp388.c]: It looks like you editor settings might be off; please set tab=8

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thank you for the fix. I'm sorry for the delay in my review.

Please rebase and address the compliance errors. There are a lot of lines with leading spaces that need to be changed to tabs

**Pull request closed without merging**